evap/student/views.py

```diff
@@ -10,38 +10,41 @@
 
 
 @participant_required
 def index(request):
-    # retrieve all evaluations, where the user is a participant and that are not new
-    evaluations = list(set(Evaluation.objects.filter(participants=request.user).exclude(state="new")))
+    # retrieve all courses which have evaluations that are not in state "new" and in which the user participates
+    courses = Course.objects.filter(
+        evaluations__participants=request.user,
+        evaluations__state__in=['prepared','editor_approved','approved','in_evaluation','evaluated','reviewed','published']
+    )
+    # retrieve all evaluations which the user can see that are not new
+    evaluations = list(set([evaluation for course in courses for evaluation in course.evaluations.all() if evaluation.state != "new" and evaluation.can_user_see_evaluation(request.user)]))
     for evaluation in evaluations:
-        evaluation.distribution = calculate_average_distribution(evaluation)
-        evaluation.avg_grade = distribution_to_grade(evaluation.distribution)
-
-    voted_evaluations = list(set(Evaluation.objects.filter(voters=request.user)))
-    due_evaluations = list(set(Evaluation.objects.filter(participants=request.user, state='in_evaluation').exclude(voters=request.user)))
-
-    # due evaluations come first, then everything else in chronological order
-    # some states are handled as a group because they appear the same to students
-    def sorter(evaluation):
-        return (
-            evaluation not in due_evaluations,
-            evaluation.state not in ['prepared', 'editor_approved', 'approved'],
-            evaluation.state != 'in_evaluation',
-            evaluation.state not in ['evaluated', 'reviewed'],
-            evaluation.full_name
-        )
-    evaluations.sort(key=sorter)
+        if not evaluation.is_single_result:
+            evaluation.distribution = calculate_average_distribution(evaluation)
+            evaluation.avg_grade = distribution_to_grade(evaluation.distribution)
+        else:
+            evaluation.single_result_rating_result = get_single_result_rating_result(evaluation)
+        evaluation.participates_in = request.user in evaluation.participants.all()
+        evaluation.voted_for = request.user in evaluation.voters.all()
+        if evaluation.course.evaluations.count() > 1:
+            if any(evaluation.state != "published" for evaluation in evaluation.course.evaluations.all()):
+                evaluation.course.not_all_evaluations_are_published = True
+            evaluation.course.distribution = calculate_average_course_distribution(evaluation.course)
+            evaluation.course.avg_grade = distribution_to_grade(evaluation.course.distribution)
+            evaluation.course_weight_sum = sum(evaluation.weight for evaluation in evaluation.course.evaluations.all())
+    evaluations.sort(key=lambda evaluation: evaluation.full_name)  # evaluations must be sorted for regrouping them in the template
```

Comment:
Feels like duplication here again.

Reply:
only the grade calculation or the sorting as well? i now moved the grade calculation.